### DISCUSSION TITLE
0.5->0.6 migration advice?

### DISCUSSION THREAD

**Discussion by mock1328** [Relm4/Relm4#659] [Q&A] [answered]
For study purpose, I tried upgrading the package that I use, regreet, onto newer `gtk-rs`/`relm4` version. However, everything above `0.5` fails to compile, [this](https://github.com/rharish101/ReGreet/blob/1b3b53278d9df5b9d48d7a9f467e199856b8c313/src/gui/component.rs#L137) `view!` components, to be specific. It can't find the `template_child` elements, even though vscode manages to find them correctly into the `templates.rs` file, and these have correct `#[name]` attribute set up:

(example of error for `user_toggle`
```
cannot find value `user_toggle` in this scope
not found in this scope
```

But I guess this was fixed in [0.6.2](https://github.com/Relm4/Relm4/blob/main/CHANGES.md#062---2023-8-27), so when upgrading to `0.7` or higher versions, some of compilation errors go away, but some related to `connect_clicked` are not:

e.g. for this code:

```rust
                    connect_activate[
                        sender, usernames_box, username_entry, sessions_box, session_entry
                    ] => move |this| {
                        sender.input(Self::Input::Login {
                            input: this.text().to_string(),
                            info: UserSessInfo::extract(
                                &usernames_box, &username_entry, &sessions_box, &session_entry
                            ),
                        })
                    }
```

I have an error:

```
cannot find value `usernames_box` in this scope
component.rs(229, 34): the binding `usernames_box` is available in a different scope in the same function
```

Thourgh trial-and-error I've managed to fix (?) it by making these references with `&`:

```rust
                    connect_activate[
                        sender, &usernames_box, &username_entry, &sessions_box, &session_entry
                    ] => move |this| {
```

However, now it errors on `move` statement:

```
used binding `usernames_box` isn't initialized
`usernames_box` used here but it isn't initialized
...
```

I've asked the same question in [gtk-rs](https://github.com/gtk-rs/gtk4-rs/discussions/1799) repo, and they say it's probably related to my relm4 usage, rather than gtk-rs breakage. The [migration guide](https://relm4.org/book/stable/migrations/0_5_to_0_6.html), changelog, changes in examples between `0.5.1` and `0.6.0` tags are not giving useful information, so I'm a bit stuck...

**Comment by AaronErhardt** [comment, marked as answer]
I think in 0.5 template children accidentally "tainted" the scope of the entire view macro (because they were globally available, which made it impossible to refer to the children when using the same template twice). Now you should be able to access it like you would do in regular Rust: `parent.child_name`.

In you case, that would mean something like this:

```rust
                    connect_activate[
                        sender, usernames_box = ui.usernames_box.clone(), username_entry, sessions_box, session_entry
                    ]
```

**Reply by mock1328** [reply]
Thanks for advice! Works like a charm :)
